web/src/message_store.ts

```diff
@@ -394,16 +400,179 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
+};
+
+// stream_id -> topic -> set of TopicLinks
+const topic_links_by_from = new Map<number, Map<string, Set<TopicLink>>>();
+const topic_links_by_to = new Map<number, Map<string, Set<TopicLink>>>();
+
+export function topic_links_from_narrow(
+    stream_id: number,
+    topic: string,
+): {
+    display_html: string;
+    url: string;
+}[] {
+    const topic_links = topic_links_by_from.get(stream_id)?.get(topic)?.values() ?? [];
+    const link_content = topic_links.map((topic_link) =>
+        topic_link_util.get_topic_link_content_with_stream_id(
+            topic_link.to_stream_id,
+            topic_link.to_topic,
+            topic_link.to_message_id?.toString(),
+        ),
+    );
+    return [...link_content];
+}
+
+export function topic_links_to_narrow(
+    stream_id: number,
+    topic: string,
+): {
+    display_html: string;
+    url: string;
+}[] {
+    const topic_links = topic_links_by_to.get(stream_id)?.get(topic)?.values() ?? [];
+    return [...topic_links].map((topic_link) =>
+        topic_link_util.get_topic_link_content_with_stream_id(
+            topic_link.from_stream_id,
+            topic_link.from_topic,
+            topic_link.from_message_id.toString(),
+        ),
+    );
+}
+
+function filter_set<T>(set: Set<T>, filter: (item: T) => boolean): Set<T> {
+    return new Set([...set].filter((item) => filter(item)));
+}
+
+function remove_topic_links(stream_id: number, topic: string, message_id: number): void {
+    const stream_links_from_narrow = topic_links_by_from.get(stream_id);
+    if (stream_links_from_narrow?.has(topic)) {
+        const filtered_links = filter_set(
+            stream_links_from_narrow.get(topic)!,
+            (link) => link.from_message_id !== message_id,
+        );
+        stream_links_from_narrow.set(topic, filtered_links);
+    }
+
+    const stream_links_to_narrow = topic_links_by_to.get(stream_id);
+    if (stream_links_to_narrow?.has(topic)) {
+        const filtered_links = filter_set(
+            stream_links_to_narrow.get(topic)!,
+            (link) => link.to_message_id !== message_id,
+        );
+        stream_links_to_narrow.set(topic, new Set(filtered_links));
+    }
+}
+
+function set_topic_links(
+    link_map: Map<number, Map<string, Set<TopicLink>>>,
+    topic_link: TopicLink,
+    stream_id: number,
+    topic: string,
+): void {
+    const stream_links = link_map.get(stream_id) ?? new Map<string, Set<TopicLink>>();
+    const topic_links = stream_links.get(topic) ?? new Set<TopicLink>();
+    topic_links.add(topic_link);
+    link_map.set(stream_id, stream_links);
+    stream_links.set(topic, topic_links);
+}
+
+function save_topic_links(message: Message): void {
+    if (message.type === "private") {
+        return;
+    }
+
+    const stream_topic_regex =
+        /#narrow\/channel\/(\d+)-[^/]*(?:\/topic\/([^/]*)(?:\/near\/(\d+))?)?/;
+    for (const link of $(message.content).children("a")) {
+        const match = stream_topic_regex.exec(link.href);
```

Comment:
The JQuery and regex feel janky, but I can't think of a better way to do it given that we don't already have the data stored on these messages. Going forward we can store the data in data attrs, though I wasn't sure if it was worth it to add that and have two blocks of code handling this different ways.